Implement Observable#last

maraisr opened the pull request:
#### 39c23277bab79fb17fa560e50f4adbdee7d13020
<pre>
Implement Observable#last
<a href="https://bugs.webkit.org/show_bug.cgi?id=277512">https://bugs.webkit.org/show_bug.cgi?id=277512</a>

Reviewed by Darin Adler.

This change introduces the `.last` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.
This is required by the specification <a href="https://wicg.github.io/observable/#dom-observable-last">https://wicg.github.io/observable/#dom-observable-last</a>

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.js:
  AbortErrors&apos; description text is not normative, as such do not assert
  against it.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverLast.cpp: Added.
(WebCore::createInternalObserverOperatorLast):
* Source/WebCore/dom/InternalObserverLast.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::last):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
  Exposes `Promise&lt;any&gt; last(options)` as a promise-returning operator.

Canonical link: <a href="https://commits.webkit.org/285542@main">https://commits.webkit.org/285542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ec17b43773ec712878a319115928c540f925714

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72220 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25008 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76388 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74335 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23251 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56929 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75287 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62189 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43444 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19663 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21779 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20024 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78065 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19183 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65393 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16508 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62213 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64656 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16086 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12875 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6519 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47439 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2223 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48508 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49796 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48251 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+void createInternalObserverOperatorLast(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>& promise)
+{
+    if (UNLIKELY(options.signal)) {
+        Ref<AbortSignal> signal = *options.signal;
```
Jut `Ref` is sufficient, don’t need to write `Ref<AbortSignal>`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+        auto lastValue = std::exchange(m_lastValue, std::nullopt);
+
+        if (UNLIKELY(!lastValue))
+            return m_promise->reject(Exception { ExceptionCode::RangeError,  "No values in Observable"_s });
```
I think we need to copy `m_promise` into a local `Ref`.

Extra space here after the comma.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+        m_promise->resolve<IDLAny>(*lastValue);
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final
+    {
+    }
```
Do these have to be overridden?

maraisr replied:
They are virtual in the base class, and could not get the compiler to compile without declaring those here. Perhaps there is a better way?

https://github.com/WebKit/WebKit/blob/a38f1eae05cc9330d30874f98bc9af67b72dbfe7/Source/WebCore/dom/InternalObserver.h#L54-L55

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    void error(JSC::JSValue value) final
+    {
+        m_promise->reject<IDLAny>(value);
```
I think we need to copy `m_promise` into a local `Ref` in case calling reject has a side effect that ends up deleting the observer. Something like this: `Ref { m_promise }->reject<IDLAny>(value)`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+        if (UNLIKELY(!lastValue))
+            return m_promise->reject(Exception { ExceptionCode::RangeError,  "No values in Observable"_s });
+
+        m_promise->resolve<IDLAny>(*lastValue);
```
I think we need to copy `m_promise` into a local `Ref`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.h:
```diff
@@ -0,0 +1,39 @@
+class ScriptExecutionContext;
+struct SubscribeOptions;
+
+void createInternalObserverOperatorLast(ScriptExecutionContext&, Ref<Observable>, SubscribeOptions, Ref<DeferredPromise>&);
```
Usually we’d make the argument be an `Observable&` instead of a `Ref<Observable>`. But we require that the caller makes a `Ref` anyway, so I guess we can just leave it like this.

That last argument should be `Ref<DeferredPromise>&&`, an rvalue reference, not an lvalue reference.

darinadler commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -119,6 +120,11 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
 
+void Observable::last(ScriptExecutionContext& context, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    return createInternalObserverOperatorLast(context, *this, options, promise);
```
Should be `WTFMove(promise)`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+        });
+    }
+
+    auto observer = InternalObserverLast::create(context, promise);
```
Should be `WTFMove(promise)`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+    {
+    }
+
+    InternalObserverLast(ScriptExecutionContext& context, Ref<DeferredPromise>& promise)
```
Should be `Ref<DeferredPromise>&&`, an rvalue reference.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    InternalObserverLast(ScriptExecutionContext& context, Ref<DeferredPromise>& promise)
+        : InternalObserver(context)
+        , m_promise(promise)
```
Should be `WTFMove(promise)`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+
+class InternalObserverLast final : public InternalObserver {
+public:
+    static Ref<InternalObserverLast> create(ScriptExecutionContext& context, Ref<DeferredPromise>& promise)
```
Should be `Ref<DeferredPromise>&&`, an rvalue reference.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+public:
+    static Ref<InternalObserverLast> create(ScriptExecutionContext& context, Ref<DeferredPromise>& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverLast(context, promise));
```
Should be `WTFMove(promise)`.

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    void error(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->reject<IDLAny>(value);
```
I think @cdumez has been writing `Ref { m_promise }.get()->` now that I see this. I guess we should pick a style. Another option is to add a `protectedPromise()` function and call it each time.

maraisr commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.js:
```diff
@@ -65,7 +65,6 @@ promise_test(async () => {
       "Promise rejects with a DOMException for abortion");
   assert_equals(rejection.name, "AbortError",
       "Rejected with 'AbortError' DOMException");
-  assert_equals(rejection.message, "signal is aborted without reason");
```
https://github.com/web-platform-tests/wpt/pull/48717, https://github.com/WICG/observable/pull/179#discussion_r1807557771

maraisr committed:
Implement Observable#last
https://bugs.webkit.org/show_bug.cgi?id=277512

Reviewed by Darin Adler.

This change introduces the `.last` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.
This is required by the specification https://wicg.github.io/observable/#dom-observable-last

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.js:
  AbortErrors' description text is not normative, as such do not assert
  against it.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverLast.cpp: Added.
(WebCore::createInternalObserverOperatorLast):
* Source/WebCore/dom/InternalObserverLast.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::last):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
  Exposes `Promise<any> last(options)` as a promise-returning operator.

Canonical link: https://commits.webkit.org/285542@main
